### PR TITLE
Learning the mapping between gaussian maps and nbody + baryon + ia ones (~training the vae)

### PR DESCRIPTION
**Pull request by Justinezgh**
This PR aims to keep track of all VAE training experiments.

### CODE REVIEW

**Comment by Justinezgh** [comment]
The script used for the first attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/u/Justinezgh/vae/scripts/vae/train_vae.py). I used the following configuration: 

```python 
lr_rate = 1e-2
total_steps = 50000
batch = 49999
```

Here are some results: 
image
image
image
image
image

In this [notebook](https://github.com/Justinezgh/Learn2Map/blob/u/Justinezgh/vae/notebooks/vae/dev/explicit_inference_with_learned_trasnfo.ipynb), I performed the inference by building an explicit learned model that I sampled using NUTS (I guess the best would be to do some SBI starting from the Gaussian maps etc but it was faster this way). Here are the results: 
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/24eebf9bfbdf51b9f490bbee1d42b362ba35810c/scripts/vae/train_vae.py). The main change is that I used a small unet as decoder.

Here are some results: 
image
image
image
image
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/2099fa46f4bb594a9f625449889ead0d7ea93188/scripts/vae/train_vae.py). The main change is that I add a weight = 0 on the dkl term of elbo loss

Here are some results
image
image
image
image
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/1b05e68325a3dddeeab9d55c974a9e266589a5fe/scripts/vae/train_vae.py). The main change is that there is no p(z|x).  Here are some results: 
image
image

The encoder seems to be the main pb here. @b-remy you have an idea of what we can do to make it perform better? Do you still think it is an up sampling pb?

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/43ba6c45193dc3c94c7f0c15ef2b6364a2709270/scripts/vae/train_vae.py). The main change is that I use the same simple convolutional layer as the decoder for the encoder.  Here are some results: 

image
image

File: learn2map/models/unet_model.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+import types
+from typing import Mapping, Optional, Sequence, Union
+
+def check_length(length, value, name):
+  if len(value) != length:
+    raise ValueError(f"`{name}` must be of length 4 not {len(value)}")
+
+def upsample(image):
+    b, h, w, c = image.shape
+    upsampled_image = ji.resize(image, [b, 2*h, 2*w, c], 'nearest')
```
Comment (b-remy) [inline learn2map/models/unet_model.py:16]: If there is a persistent pattern in the output of the unet, this is sometimes due to the upsampling operation (here nearest-neighbor [interpolation](https://jax.readthedocs.io/en/latest/_autosummary/jax.image.resize.html)). We may want to try other upsampling operations, for instance, `bilinear` or `cubic` interpolation

Comment (Justinezgh) [reply learn2map/models/unet_model.py:16]: okk thanks :)

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/f13a3f93f99b439ca710cd4ee36a07822f63b777/scripts/vae/train_vae.py). The main change is that I reduce the observational noise.  Here are some results: 
image
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/fdb77e46e764d792d199209e3bb663eeb68a110e/scripts/vae/train_vae.py). The main change is that I put back the UResnet.  Here are some results: 
image
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/0ec2dc360e83a6743a2670bff79b7ba9ef248625/scripts/vae/train_vae.py). The main change is that I added some observational noise  Here are some results: 
image
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/cfe9b5089f9beeb4385d9ddab7d305c1a6efd93b/scripts/vae/train_vae.py). The main change is that I added realistic observational noise  Here are some results: 
image
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/db1274f8711eeca4694a8c0d7c1a260a28a8e8be/scripts/vae/train_vae.py). The main change is that I put back the posterior p(z|x). Here are some results:
image
image
image

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/a57279af4e98bd972a359ac211165507291a88af/scripts/vae/train_vae.py). The main change is that I put back the dkl. Here are some results:
image
image
image
image

**Comment by Justinezgh** [comment]
I trained for 44_000 iterations and change the learning rate schedule.
image
image
image

**Comment by b-remy** [comment]
I don't think the latent space samples should be noisy, and I expect them to have the same power spectrum as the prior in the DKL. 

x-xpred should be noise only indeed :), so it should have a zero mean and same std as the observation noise, if not it is because there are errors in the prediction.

One thing I noticed in the last figure above is that the power spectrum of the observations is highly dominated by the noise, as if a very few scales of the map are present after noise corruption. It makes me wonder if we wouldn't want to work with larger fields

**Comment by Justinezgh** [comment]
If I average over 300 latent maps like this: 
```python 
out_encoder, _ = encoder_eval.apply(vae_params, state[0], m_data_proj_noisy[0])
sample_encoder = posterior_z(out_encoder.squeeze()).sample(300,seed=jax.random.PRNGKey(0))
mean_sample_encoder = jnp.mean(sample_encoder, axis = 0)
out_vae, _ = decoder_eval.apply(vae_params, state[1], mean_sample_encoder)
sample_vae = posterior_x(out_vae.squeeze()).sample(jax.random.PRNGKey(10))
``` 
image
image
image
image

**Comment by b-remy** [comment]
The prower spectrum for the prediction looks great! What happens when the decoder takes as input a Gaussian map?

**Comment by Justinezgh** [comment]
```python 
out_vae, _ = decoder_eval.apply(vae_params, state[1], projection_gaussian)
sample_vae = posterior_x(out_vae.squeeze()).sample(jax.random.PRNGKey(10))
```

image
image

(and as a reminder, the PS of Gaussian and Nbody are supposed to be the same)

**Comment by Justinezgh** [comment]
If we centralize: 
```python 
out_vae, _ = decoder_eval.apply(vae_params, state[1],projection_gaussian - projection_gaussian.mean())
sample_vae = posterior_x(out_vae.squeeze()).sample(jax.random.PRNGKey(10))
```
image
image

**Comment by Justinezgh** [comment]
Todo: 
- [ ] @b-remy check that Gaussian maps are ok 
- [ ] @b-remy put back bacthnorm
- [ ] @Justinezgh schedule dkl & schedule nll

**Comment by Justinezgh** [comment]
The script used for this attempt of training the VAE is [here](https://github.com/Justinezgh/Learn2Map/blob/f8643ec2cb41fe022501450b1b297b78f6085bdd/scripts/vae/train_vae.py). The main change is that I put scheduler on the dkl. Here are some results:
image
image
image
image
when I average over 300 z:
image
image
image
image

**Comment by Justinezgh** [comment]
ok we should indeed decrease the noise 
image

**Comment by Justinezgh** [comment]
We added the batch norm 
image

image
image
image
image

Testing decoder: 
image
image

**Comment by Justinezgh** [comment]
Issues: 
- [ ] residuals not 0 centered
- [ ] normalize latent space to avoid this
image
image

**Comment by Justinezgh** [comment]
I added a linearity and additional convolution in decoder.
image
image
image
image
image

testing decoder:
image
image

**Comment by b-remy** [comment]
After adding the convolutional layer to the decoder, I think it happens what I was expecting, i.e. the encoder has more freedom to encode in a latent space which is closer to a GRF, while the autoencoder still being able to recover a good reconstruction. So I think we can still try to increase the decoder flexibility a little bit, and see if it improves both the reconstruction and the latent space distribution :)

One thing I don't understand though: although the _true gaussian samples_ generated on the sphere and projected to a 80x80 map seem to have the correct power spectrum, they don't have the same range of values compared to the samples generated from our GRF prior. Therefore the range of values seen by the decoder at inference time will not be the same as seen from training, which can lead to the weird patterns in the last figure.

**Comment by Justinezgh** [comment]
Do you have an architecture in mind for the decoder? or I just add an additional convolution like that: 

```python 
class ConvDecoder(hk.Module):
    def __init__(self):
        super().__init__()
    def __call__(self, x):
        residual = hk.Conv2D(1, 3, 1)(x)
        residual = jax.nn.leaky_relu(residual)
        residual = hk.Conv2D(1, 3, 1)(residual) 
        residual = jax.nn.leaky_relu(residual)
        residual = hk.Conv2D(1, 3, 1)(residual)
        return (residual + x).squeeze()
```

I guess at the end we want to take the Gaussian map that we build from Cosmogrid simulations. But yes in the notebook you send I use the gaussian maps from the same prior I have in the latent space of the vae and the results are not that good for the moment:
image
These results corresponds to [this experiment](https://github.com/Justinezgh/Learn2Map/pull/3#issuecomment-2162562407)

And here is the prediction from the decoder starting from a gaussian map sampled from the prior: 
image
image

**Comment by Justinezgh** [comment]
I added an additional convolutional layer in the decoder. Here are some results:

## Training loss
image

## Checking VAE
```python 
out_encoder, _ = encoder_eval.apply(vae_params, state[0], m_data_proj_noisy[0])
sample_encoder = posterior_z(out_encoder.squeeze()).sample(
    seed=jax.random.PRNGKey(0)
)
out_vae, _ = decoder_eval.apply(vae_params, state[1], sample_encoder)
sample_vae = posterior_x(out_vae.squeeze()).sample(jax.random.PRNGKey(10))
```
image

 ⚠️ ⬆️ RESIDUALS ARE CENTERED 🥳

image
image
image

## Checking VAE while averaging over z
```python 
out_encoder, _ = encoder_eval.apply(vae_params, state[0], m_data_proj_noisy[0])
sample_encoder = posterior_z(out_encoder.squeeze()).sample(300,seed=jax.random.PRNGKey(0))
mean_sample_encoder = jnp.mean(sample_encoder, axis = 0)
out_vae, _ = decoder_eval.apply(vae_params, state[1], mean_sample_encoder)
sample_vae = posterior_x(out_vae.squeeze()).sample(jax.random.PRNGKey(10))
```
image
image
image
image

## Checking decoding with Cosmogrid Gaussian map
```python 
out_vae, _ = decoder_eval.apply(vae_params, state[1],projection_gaussian)
sample_vae = posterior_x(out_vae.squeeze()).sample(jax.random.PRNGKey(10))
```
image
image

## Checking decoding with prior Gaussian map
```python 
out_vae, _ = decoder_eval.apply(vae_params, state[1],gaussian_map_prior)
sample_vae = posterior_x(out_vae.squeeze()).sample(jax.random.PRNGKey(10))
```
image
image

**Comment by Justinezgh** [comment]
Explicit inference results: 

```python
def learned_forward_model():
    Om = numpyro.sample("Om", dist.Uniform(0.1,0.5))
    s8 = numpyro.sample("s8", dist.Uniform(0.4,1.4))
    h = numpyro.sample("h", dist.Uniform(0.64, 0.82))
    Ob = numpyro.sample("Ob", dist.Uniform(0.03,0.06))
    ns = numpyro.sample("ns", dist.Uniform(0.87, 1.07))
    w0 = numpyro.sample("w0", dist.Uniform(-2.0, -0.333))
    
    z = numpyro.sample(
        "z", dist.MultivariateNormal(loc=jnp.zeros((N, N)), precision_matrix=jnp.eye(N))
    )
    
    cosmo = jc.Planck15(
        Omega_c=Om - Ob,
        Omega_b=Ob,
        h=h,
        n_s=ns,
        sigma8=s8,
        Omega_k=0.0,
        w0=w0,
        wa=0.0,
    )
    
    P = partial(Pk_fn, cosmo=cosmo, a_ai=None)
    power_map = make_power_map(P, N, map_size_rad)
    
    gaussian_map = jnp.fft.ifft2(jnp.fft.fft2(z) * jnp.sqrt(power_map)).real
    gaussian_map = gaussian_map - gaussian_map.mean()

    out_vae = predicted_nbody(gaussian_map.reshape([-1,N,N]))
    
    x = numpyro.sample(
        "x",
        dist.Independent(
            dist.Normal(out_vae.squeeze(), sigma_e / jnp.sqrt(gal_per_arcmin2 * pix_area)), 2
        ),
    )
    
    return x
```
image

**Comment by Justinezgh** [comment]
Implicit inference results with mse compression: 
image

image

These results have to be compraed with  [these results](https://github.com/Justinezgh/Learn2Map/blob/f99f046bf94457ef131ff3d8a62f7b4bba08c451/notebooks/sbi/nle_mse_nbody_baryon_ia.ipynb). But note that we are note using the same galaxy density.

**Comment by b-remy** [comment]
The VAE results are much better! Both for the reconstruction and for the latent space :-) Indeed the residues are now centered (even it seems to remain a little bit of structures) and the decoder behaves well when we input a Gaussian map from the prior with the fiducial cosmology (at the power spectrum level).

Regarding your questions here is what I think:
- validating the N-body beyond the power spectrum might be interesting. This could be done with higher order statistics (Minkowsky functionals, Peak Counts, etc., there are many)
- The power spectrum might be enough for the latent space as we expect it to be Gaussian. I think it is more important to check that these maps are Gaussian with the correct power spectrum than checking that the maps are sampled from the correct distribution. And even more important that the decoder behaves well when we input a Gaussian map (as done at inference time)
- This I don't know. Maybe. I think the most important thing is to get the correct power spectrum for the map with noise since the log-likelihood will model the noise uncertainty

Regarding the inference results, we are not there yet ^^
- Yes we need the same noise model during the training and at inference time.
- I am thinking about what the decoder sees at inference and during the training... and at inference time it sees GRF with varying cosmologies, while only the fiducial cosmology during training. Isn't a potential source of bias? Since we use the theory to set the power spectrum of the Gaussian prior in latent space, could we try sample the cosmology during the training, still training to learn the function f(GRF) = N-body, but with GRF of different cosmologies. Alternative training could be
  - using a cosmo dependent prior $`p(z|\theta)`$ and sampling theta during training
  - same thing as above but also conditioning the encoder and the decoder on $`\theta`$
 
This way, the decoder would be informed of the cosmology of its input, and be less bias when using different cosmologies at inference time. What do you think?

**Comment by b-remy** [comment]
Oh and I don't think we should use this for the forward model, right?
```python
gaussian_map = gaussian_map - gaussian_map.mean()
```

**Comment by Justinezgh** [comment]
The main changes are the power spectrum normalization in the decoder, we centralized the maps and we change the dataset now it is Nbody without baryon and without IA. 
# Training loss 
image

# Checking VAE
image
image
image

# Checking latent space 
image

# Checking decoding 
## Cosmogrid Gaussian 
image
image
image


# Gaussian prior
image
image
image

# Different cosmo 
image
image

**Comment by b-remy** [comment]
@Justinezgh , note that I just made a commit in this branch to add the pixel window in the prior, so don't forget to git pull before designing your next experiment! See #10 for further explainations

**Comment by Justinezgh** [comment]
Cutting ell in Gaussian prior produce NaNs during training + now the training is super slow ~ 6s/it -> 90h in total

**Comment by b-remy** [comment]
> Cutting ell in Gaussian prior produce NaNs during training

On my side the scale cut in the Gaussian prior does not produce nans.

> now the training is super slow ~ 6s/it -> 90h in total

Indeed, I've noticed two sources of slowdown.
1. Renormalizing the decoder output with its input power spectrum generate a x2 slowdown. Each iteration will be slower, but it is still tractable
2. The scale cut in the augmentation function, which is done on the fly, is insanely slow... I see three options:
a. we find an efficient way to do this in tensorflow
b. we precompute scale cut in the dataset, which might be the cleanest way :-) but less flexible if we want to change the scale cut
c. we write the augmentation fully in jax -> I did it, it's working as fine as before and the scale cut does not increase at all the training pace
